### PR TITLE
belinda-nestjs_01-23_25_26_27-product_&_auth_modules_creation

### LINKED ISSUE #23: Implement a product data transfer object (DTO)
As a backend developer, I will want to have a product DTO to allow for easier and simplified data transfer.

**Is your feature request related to a problem? Please describe.**\
I'm always frustrated when I am having issues trying to access data in the backend.

**Describe the solution you'd like**\
A product DTO that will allow for easy access to the data that I need.

**Include link(s) to relevant tutorial(s)**\
https://www.youtube.com/watch?v=6-mBxManIBw&list=PL_cUvD4qzbkw-phjGK2qq0nQiG6gw1cKK&index=6

**Describe alternatives you've considered**\
No alternatives

**Is this issue related to any other issues?**\
"This issue relates to issue #20 

**Additional context**\

Image

Image

**Story Points**\
4-8 story points

### LINKED ISSUE #25: Product Controller & Service Class Creation
As a backend developer I want to ensure the routes and logic are in place for the clothing products stored and shared by Belinda's Closet. [see wiki API Endpoints](https://github.com/SeattleColleges/belindas-closet-nestjs/wiki/API-Endpoints) & [Nest.js YT Crash Course Tutorials](https://www.youtube.com/watch?v=xzu3QXwo1BU&list=PL_cUvD4qzbkw-phjGK2qq0nQiG6gw1cKK&pp=iAQB) to get started.

**Describe the solution you'd like**\
I want to provide routes to the frontend to perform CRUD operations regarding clothing products.

**Story Points**\
6 to 8 story points

**NSC-Events Team Members Tasked with Backend Dev Work**\
@Exochos @Ali-7s

### LINKED ISSUE #26: Product Schema Creation
As a backend developer I want to have a clearly defined data structure that contains all the fields necessary for the apps features and internal operations.
[see tutorials](https://www.youtube.com/watch?v=xzu3QXwo1BU)
**Is your feature request related to a problem? Please describe.**\
This will serve as a standard for others working across the Belinda's Closet Dev Team. Understanding the fields and structure of the event will be necessary for:

- UI Display
- Input fields
- etc.

**Additional context**\
In the early stages this may undergo significant change. _**IF**_ making any changes to the schema the following will need to be done:
- Communication with Lead Dev
- DTO updates

> Searching for the schema's use in the code base (CTRL + SHIFT + F) will quickly let you know which portions of the code will be affected when making changes to the schema.

**Story Points**\
2 - 3 Story Points

### LINKED ISSUE #27: Auth Service & Controller Creation
As a backend developer I will use passport to setup authentication and authorization. (see Wiki for details)

**Describe the solution you'd like**\
The following will need to be done:

- [x] Create login.dto.ts
- [x] Create signup.dto.ts
- [x] Create user schema user.schema.ts (for login purposes)
- [x] implement jwt strategy

Relevant Tutorial Videos
[see wiki resources](https://www.youtube.com/watch?v=xzu3QXwo1BU&list=PL_cUvD4qzbkw-phjGK2qq0nQiG6gw1cKK). VIdeos 10, 11 and 12

**Story Points**\
8 to 12 story points

### PR DESCRIPTION
**Pull request by tinpham5614**
Resolves #23, resolves #25, resolves #26, resolves #27

This PR will update the modules to match with the NSC Event NestJS.
1. Modified **products module** by adding product DTO and product schema
2. Modified **user schema**
3. Added **auth module** with controller, service, and DTO
4. Implemented **JWT strategy** (sign up and login will return a token)

To manually test on Postman:
1. Create a **.env** file in the root directory and add the variables below:
```
MONGODB_URI=<This is your MongoBD connect string>
NEXTAUTH_URL=http://localhost:3000
NEXTAUTH_SECRET=secret
JWT_SECRET=secret
JWT_EXPIRES="1d"
PRODUCT_KEY=secret
```
2. This is a **product** example in JSON:
```
{
    "productId": "ABC123",
    "productType": [
        "Shoes"
    ],
    "gender": [
        "Male"
    ],
    "productShoeSize": [
        "8"
    ],
    "productSize": [
        "M"
    ],
    "productSizePantsWaist": [
        "32"
    ],
    "productSizePantsInseam": [
        "30"
    ],
    "productDescriptionOptional": "A new style Shoes",
    "productImage": "shoes.jpg"
}
```

Below are screenshots by manual test using Postman:

**1. Add Product**
Add Product

**2. Get All Products**
Get All Product

**3. Get A Product by ID**
Get A Product By ID

**4. Update Product by ID**
Update Product by ID

**5. Sign Up**
Sign Up
Sign Up token

**6. Login **
Log In
Log In token

### CODE REVIEW

File: src/products/controllers/products/products.controller.ts [hunk 1]
```diff
@@ -5,52 +5,29 @@ import {
...
-      @Body('gender') gender: [],
-      @Body('productShoeSize') productShoeSize: [],
-      @Body('productSize') productSize: [],
-      @Body('productSizePantsWaist') productSizePantsWaist: [],
-      @Body('productSizePantsInseam') productSizePantsInseam: [],
-      @Body('productDescriptionOptional') productDescriptionOptional: string,
-      @Body('productImage') productImage: string,
-
-      ) {
+    async createProduct(@Body() productDto: CreateProductDto) {
```
Comment (Robel-003) [inline src/products/controllers/products/products.controller.ts:19]: Just for postman purposes, having a template of what fields are required would be helpful when testing the endpoints. 
ex: 
{
    "productId": "ABC123",
    "productType": [
        "Shoes"
    ],
...
}

Comment (Robel-003) [reply src/products/controllers/products/products.controller.ts:19]: Other than that, it looks good! I tested the endpoints in postman and they're all working

Comment (tinpham5614) [reply src/products/controllers/products/products.controller.ts:19]: Thank you for your suggestion. 👍

File: src/products/controllers/products/products.controller.ts [hunk 2]
```diff
@@ -5,52 +5,29 @@ import {
...
     Res,
+    Put,
   } from '@nestjs/common';
-  import { ProductsService } from 'src/products/services/products/products.service';
-
-  
+import { ProductsService } from 'src/products/services/products/products.service';
+import { CreateProductDto } from 'src/products/dto/create-product.dto';
+import { UpdateProductDto } from 'src/products/dto/update-product.dto';
   @Controller('products')
```
Comment (theGaryLarson) [inline src/products/controllers/products/products.controller.ts:14]: My fault, I didn't update the Wiki correctly. I missed a couple things regarding the routes. We want to ensure routes are standard across our repositories.

I updated the Wiki with the correct endpoints. Can you update this here as well? Thank you 

## Product Controller (`/products`)

> **Note**: These endpoints need created using Controller & Service.

- `GET /products`: Fetch all products.
- `GET /products/find/:id`: Fetch a specific product by its ID.
- `POST /products/new`: Create a new product.
- `PUT /products/update/:id`: Update an existing product by its ID.
- `DELETE /products/remove/:id`: Delete a product by its ID.

## User Controller (`/user`)

> **Note**: These endpoints need created using Controller & Service.

- `POST /user/new`: Add a new user.
- `GET /user`: Fetch all users.
- `GET /user/find/:id`: Fetch a specific user by their ID.
- `GET /user/email/:email`: Fetch a user by their email.
- `PUT /user/update/:id`: Update an existing user by their ID.
- `DELETE /user/remove/:id1: Remove an existing user by their ID.

**Review by theGaryLarson — DISMISSED** [review]
Looks awesome Tin! Thank you. I made a couple mistakes when defining the endpoints. We want them to be the same across repos. I updated the wiki. My apologies.. Could you please update the controllers.

Also, I'd like to get together with you and discuss the registration key and next tasks: Authentication / UseGuards() and 
## Product Controller (`/products`)

> **Note**: These endpoints need created using Controller & Service.

- `GET /products`: Fetch all products.
- `GET /products/find/:id`: Fetch a specific product by its ID.
- `POST /products/new`: Create a new product.
- `PUT /products/update/:id`: Update an existing product by its ID.
- `DELETE /products/remove/:id`: Delete a product by its ID.

## User Controller (`/user`)

> **Note**: These endpoints need created using Controller & Service.

- `POST /user/new`: Add a new user.
- `GET /user`: Fetch all users.
- `GET /user/find/:id`: Fetch a specific user by their ID.
- `GET /user/email/:email`: Fetch a user by their email.
- `PUT /user/update/:id`: Update an existing user by their ID.
- `DELETE /user/remove/:id1: Remove an existing user by their ID.

File: src/products/controllers/products/products.controller.ts (hunk 2 above)
Comment (tinpham5614) [reply src/products/controllers/products/products.controller.ts:14]: Absolutely! I will do that. Thank you. 👍

Comment (tinpham5614) [reply src/products/controllers/products/products.controller.ts:14]: @theGaryLarson I went ahead and fixed the product API endpoints. I will let the PO/PM create issues with the user module, giving other students a chance to learn (user's controller, service, and DTO creation).

**Pull request merged**